File: problem_10.cpp
#include <chrono>
#include <iostream>
#include <vector>
#include <cstring>
#include <cmath>
#include <functional>

const long limitNumber = 2000000;

std::function<long(long, long)> sum = [] (long n, long m) 
{
    return n + m;
};

long reducePrimesUsingSieve (long n, std::function<long(long, long)> reduceFunc, long acc) 
{
    std::vector<unsigned char> sieve;
    sieve.resize(n + 1);
    memset(sieve.data(), 0, (n + 1) * sizeof(unsigned char));
    for (long i = 2; i <= int(sqrt(n)); i++) 
    {
        if (sieve[i] != 0) 
        {
            continue;
        }
        for (long j = i + i; j < n; j += i) 
        {
            sieve[j] = (unsigned short) 1;
        }
    } 
    for (long i = 2; i < n; i++) 
    {
        if (sieve[i] == 0)
        {
            acc = reduceFunc(acc, i);
        }
    }
    return acc;
}

int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();

    long sumOfPrimes = reducePrimesUsingSieve(limitNumber, sum, 0);
    
    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << sumOfPrimes << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_10.rs
use std::time::{Instant};

const LIMIT_NUMBER: usize = 2000000;

fn reduce_primes_using_sieve<F> (n: usize, f: F, mut acc: usize) -> usize
    where F: Fn(usize, usize) -> usize
{
    let mut sieve: Vec<u8> = vec![];
    sieve.resize(n + 1, 0);
    let limit = f64::sqrt(n as f64) as usize;
    for i in 2..limit 
    {
        if sieve[i] == 0 
        {
            for j in ((i + i)..n).step_by(i)
            {
                sieve[j] = 1;
            }
        }
    }
    for i in 2..n 
    {
        if sieve[i] == 0
        {
            acc = f(acc, i);
        }
    }
    return acc;
}

fn main () 
{
    let start = Instant::now();

    let result = reduce_primes_using_sieve(LIMIT_NUMBER, |acc, x| acc + x, 0);
    println!("Result: {}", result);

    let duration = start.elapsed();
    println!("Timing: {}", duration.as_nanos());
}


